src/collector.rs

```diff
@@ -234,18 +234,27 @@ pub async fn collect_links(
         }
     }
 
-    // Note: we could dispatch links to be checked as soon as we get them,
-    //       instead of building a HashSet with all links.
-    //       This optimization would speed up cases where there's
-    //       a lot of inputs and/or the inputs are large (e.g. big files).
-    let mut collected_links: HashSet<Request> = HashSet::new();
-
-    for handle in extract_links_handles {
-        let links = handle.await?;
-        collected_links.extend(links);
-    }
+    let (links_tx, links_rx) = tokio::sync::mpsc::channel(max_concurrency);
+    tokio::spawn(async move {
+        let mut collected_links = HashSet::new();
+
+        for handle in extract_links_handles {
+            // Unwrap should be fine because joining fails:
+            // * if the Task was dropped (which we don't do)
+            // * if the Task panicked. Propagating panics is correct here.
+            let requests = handle.await.unwrap();
```

Comment:
Unwrap should be fine here indeed. Can we use `expect` instead and mention the condition that didn't hold to make debugging easier just in case? E.g. `handle.await.expect("Awaiting termination of link handle failed");` or something.

Reply:
Good idea, I'll do that